pyquist/audio.py

```diff
@@ -95,14 +95,46 @@ def __new__(
         obj._sample_rate = sample_rate
         return obj
 
-    def __getitem__(self, *args, **kwargs):
-        """Ensure slices retain the sample rate."""
-        result = super().__getitem__(*args, **kwargs)
-        if isinstance(result, AudioBuffer):
-            result._sample_rate = self._sample_rate
-            assert isinstance(result, Audio)
+    def __array_function__(self, func, types, args, kwargs):
+        """Ensure sample_rate is passed along for functions like np.concatenate.
+        
+        See https://numpy.org/doc/stable/reference/arrays.classes.html#numpy.class.__array_function__"""
+
+        # Extract sample rates from all Audio instances in args
+        def extract_sample_rates(arg):
+            if isinstance(arg, Audio):
+                return [arg.sample_rate]
+            elif isinstance(arg, (list, tuple)):  # Handle nested structures
+                return [rate for elem in arg for rate in extract_sample_rates(elem)]
+            elif isinstance(arg, dict): # Handle values of dict, rarely used in Numpy, but possible
+                return [rate for elem in arg.values() for rate in extract_sample_rates(elem)]
+            else:
+                return []
+
+        sample_rates = extract_sample_rates(args)
```

Comment:
not 100% sure but it seems like this should be `extract_sample_rate` (singular) and `[extract_sample_rate(arg) for arg in args]`?